yarn-project/pxe/src/tagging/sync/sync_sender_tagging_indexes.ts

```diff
@@ -0,0 +1,121 @@
+// `docs_examples.test.ts` fail as our JSON RPC doesn't seem to allow arrays with length larger than 99. I (benesjan)
+// think this window length should be fine.
+export const WINDOW_LEN = 95;
+
+/**
+ * Syncs the highest finalized tagging index and pending tagging indexes for a given secret.
+ * @param secret - The secret that's unique for (sender, recipient, contract) tuple while the direction of
+ * sender -> recipient matters.
+ * @param app - The address of the contract that the logs are tagged for. Needs to be provided because we perform
+ * second round of siloing in this function which is necessary because kernels do it as well (they silo first field
+ * of the private log which corresponds to the tag).
+ * @remarks When syncing the indexes as sender we don't care about the log contents - we only care about the highest
+ * pending and highest finalized indexes as that guides the next index choice when sending a log. The next index choice
+ * is simply the highest pending index plus one (or finalized if pending is undefined).
+ */
+export async function syncSenderTaggingIndexes(
+  secret: DirectionalAppTaggingSecret,
+  app: AztecAddress,
+  aztecNode: AztecNode,
+  taggingDataProvider: SenderTaggingDataProvider,
+): Promise<void> {
+  // # Explanation of how syncing works
+  //
+  // When choosing an index, we select: highest pending index + 1 (or highest finalized index + 1 if no pending).
+  // If the chosen index is more than WINDOW_LEN from the highest finalized index, we throw an error. By having this
+  // hard limit we give a guarantee to a recipient that he doesn't need to look further than WINDOW_LEN ahead of the
+  // highest finalized index.
+  //
+  // This function synchronizes these two values by iteratively querying the node for a window of indexes at a time,
+  // storing all those indexes as pending, and then checking the status of each pending index to update its
+  // finalization status accordingly.
+
+  const finalizedIndex = await taggingDataProvider.getLastFinalizedIndex(secret);
+
+  let start = finalizedIndex === undefined ? 0 : finalizedIndex + 1;
+  let end = start + WINDOW_LEN;
+
+  let previousFinalizedIndex = finalizedIndex;
+  let newFinalizedIndex = undefined;
+
+  while (true) {
+    // Load and store indexes for the current window. These indexes may already exist in the database if txs using
+    // them were previously sent from this PXE. Any duplicates are handled by the tagging data provider.
+    await loadAndStoreNewTaggingIndexes(secret, app, start, end, aztecNode, taggingDataProvider);
+
+    // Retrieve all indexes within the current window from storage. For each secret-txHash pair, we only store the
+    // highest index in the data provider. This means we may not get a transaction hash even if some of its indexes
+    // fall within the window (specifically when `end` is greater than those indexes). This is fine as those tx hashes
```

Comment:
I don't understand this paragraph, what does 'its' refer to?

Reply:
This was confusing. 'its` referred to indexes within the tx.

I decided to just drop most of the comment. That info was not really needed there

https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/f1c5d80adf0136b09e8192643c07e9cc9029c5d7